### PR TITLE
Fix tuple/list multiplication and addition

### LINKED ISSUE #329: Pull request #322 breaks tuple multiplication support
Statements such as the following will be broken after this pull request
```
tup1 = (0, False, 3+1j) * 3
tup2 = tup1 * 2
```

### LINKED ISSUE #349: Tuple/List concatenation does not work
The following code:
```
a = (1,2)
b = (3,4)
c = a + b
```
is translated to:
```
program prog_test

implicit none

integer(kind=8)  :: a (0:2_8-1)
integer(kind=8), allocatable  :: c (:)
integer(kind=8)  :: b (0:2_8-1)
a = [1_8, 2_8]
b = [3_8, 4_8]
c = a + b

end program prog_test
```

### LINKED ISSUE #378: Pyccel is not compatible with textx v2.2.0
See issue https://github.com/textX/textX/issues/266

A simple fix is to change the name of the attribute `name` in `FunctionMacroStmt`

### PR DESCRIPTION
**Pull request by EmilyBourne**
Fix tuple/list concatenation (Fixes #349). Fix tuple/list multiplication (fixes #329). Fix textx 2.2.0 compatibility issues (fixes #378).

- Provide variables in the same way that they are accessed in TupleVariable function get_vars.
- Check order when defining shape of IndexedElement
- Correctly recognise tuples of type NativeGeneric() as inhomogenous
- Handle tuple/list concatenation (fixes #349)
- Move concatenation out of bugs folder
- Add tuple tests for multiplication and addition
- Rename name variable of FunctionMacroStmt class to work within textx 2.2.0 rules (fixes #378)

### CODE REVIEW

File: pyccel/ast/builtins.py [hunk 1]
```diff
@@ -194,7 +194,7 @@ def __new__(cls, *args):
     def __init__(self, *args):
         if self.stage == 'syntactic' or len(args) == 0:
             return
-        is_homogeneous = all(args[0].dtype==a.dtype for a in args[1:])
+        is_homogeneous = all(False if a.dtype is NativeGeneric() else args[0].dtype==a.dtype for a in args[1:])
         is_homogeneous = is_homogeneous and all(args[0].rank==a.rank   for a in args[1:])
```
Comment (yguclu) [inline pyccel/ast/builtins.py:198]: It seems to me that these two lines could be made more readable by combining them into a single instruction:
```
is_homogeneous = all(a.dtype is not NativeGeneric() and args[0].dtype==a.dtype and args[0].rank==a.rank for a in args[1:])
```
Does that make sense?

File: pyccel/parser/semantic.py [hunk 2]
```diff
@@ -1114,10 +1114,10 @@ def _handle_PyccelOperator(self, expr, **settings):
     def _visit_PyccelAdd(self, expr, **settings):
         args = [self._visit(a, **settings) for a in expr.args]
         if isinstance(args[0], (TupleVariable, PythonTuple, Tuple, List)):
-            expr_new = Concatenate(args, True)
-            errors.report(PYCCEL_RESTRICTION_TODO, symbol=expr,
-                bounding_box=(self._current_fst_node.lineno, self._current_fst_node.col_offset),
-                severity='fatal', blocker=self.blocking)
+            tuple_args = []
+            for a in args:
+                tuple_args += a.get_vars() if isinstance(a, TupleVariable) else a.args
```
Comment (yguclu) [inline pyccel/parser/semantic.py:1119]: Shouldn't a list comprehension be faster (and more readable) here?

Comment (EmilyBourne) [reply pyccel/parser/semantic.py:1119]: Yes, but it would require unpacking in a list comprehension which python doesn't allow

Comment (yguclu) [reply pyccel/parser/semantic.py:1119]: Right! :sweat_smile: And what about this?
```
f = lambda a: a.get_vars() if isinstance(a, TupleVariable) else a.args
tuple_args = [ai for a in args for ai in f(a)]
```
Or a little more compactly (but maybe harder to read):
```
tuple_args = [ai for a in args for ai in (a.get_vars() if isinstance(a, TupleVariable) else a.args)]
```

Comment (EmilyBourne) [reply pyccel/parser/semantic.py:1119]: Won't that produce the same output?
[f(a) for a in args] will give [[a,b,c], [d,e]]
so ai in f(a) will be [a,b,c], then [d,e]

Comment (yguclu) [reply pyccel/parser/semantic.py:1119]: Yes, but there is a double loop...

Comment (EmilyBourne) [reply pyccel/parser/semantic.py:1119]: Yes but the second loop doesn't unpack anything either. This effectively does:
```
a = [[1,2], [3,4]]
b = [ai for ai in a]
```
as f(a) returns a list

Comment (EmilyBourne) [reply pyccel/parser/semantic.py:1119]: Ah! Ok, I have understood what is going on now. You are right this works

Comment (yguclu) [reply pyccel/parser/semantic.py:1119]:
```
In [1]: args = [[1, 2, 3], [4, 5, 6]]                                                                       

In [2]: tuple_args = [ai for a in args for ai in a]                                                         

In [3]: tuple_args                                                                                          
Out[3]: [1, 2, 3, 4, 5, 6]
```

File: pyccel/parser/semantic.py [hunk 3]
```diff
@@ -1114,9 +1114,8 @@ def _handle_PyccelOperator(self, expr, **settings):
     def _visit_PyccelAdd(self, expr, **settings):
         args = [self._visit(a, **settings) for a in expr.args]
         if isinstance(args[0], (TupleVariable, PythonTuple, Tuple, List)):
-            tuple_args = []
-            for a in args:
-                tuple_args += a.get_vars() if isinstance(a, TupleVariable) else a.args
+            get_vars = lambda a: a.get_vars() if isinstance(a, TupleVariable) else a.args
+            tuple_args = [ai for a in args for ai in f(a)]
```
Comment (yguclu) [inline pyccel/parser/semantic.py:1118]: I'm afraid there is a typo: `f` >> `get_vars`.

Comment (EmilyBourne) [reply pyccel/parser/semantic.py]: Arrr, stupid mistake ^^

**Pull request merged**